marconi-chain-index/bench/BenchQueries.hs

```diff
@@ -0,0 +1,205 @@
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Aeson qualified as Aeson
+import Data.ByteString.Lazy (ByteString)
+import Data.Maybe (fromMaybe, listToMaybe)
+import Data.Text qualified as Text
+import Data.Word (Word64)
+import Database.SQLite.Simple (FromRow (fromRow), field, toRow)
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.FromField (FromField)
+import Database.SQLite.Simple.ToField (ToField)
+import Database.SQLite.Simple.ToRow (ToRow)
+import GHC.Generics (Generic)
+import Marconi.ChainIndex.Indexers (mkIndexerStream, startIndexers, utxoWorker)
+import Marconi.ChainIndex.Indexers.Utxo (StorableQuery (UtxoAddress), StorableResult (UtxoResult, getUtxoResult),
+                                         UtxoHandle, UtxoIndexer)
+import Marconi.Core.Storable (QueryInterval (QEverything))
+import Marconi.Core.Storable qualified as Storable
+import System.Environment (getEnv)
+import System.FilePath ((</>))
+import Test.Tasty.Bench (bench, bgroup, defaultMain, nfIO)
+import Text.RawString.QQ (r)
+
+data FrequencyRow a = FrequencyRow
+    { _frequencyRowValue :: !a
+    , _frequencyRowFreq  :: !Int
+    } deriving (Generic)
+
+instance (FromField a) => FromRow (FrequencyRow a) where
+    fromRow = FrequencyRow <$> field <*> field
+
+instance (ToField a) => ToRow (FrequencyRow a) where
+  toRow (FrequencyRow ad f) = toRow (ad, f)
+
+main :: IO ()
+main = do
+    -- TODO Need a better way to pass those params in the benchmark.
+    -- Look into "Custom command-line options" in the tasty-bench README.
+    nodeSocketPath <- getEnv "CARDANO_NODE_SOCKET_PATH"
+    databaseDir <- getEnv "MARCONI_DB_DIRECTORY_PATH"
+
+    indexerTVar <- atomically (newEmptyTMVar :: STM (TMVar UtxoIndexer) )
+
+    -- Run concurrently the indexing and the testing. The testing is only run once the indexing is
+    -- fully synced with the local running node.
+    race_ (runIndexerSyncing databaseDir nodeSocketPath indexerTVar) $ do
+        putStrLn "Waiting for indexer to be fully synced..."
+        waitUntilSynced databaseDir nodeSocketPath
+        putStrLn "Finished syncing!"
+        tests databaseDir indexerTVar
+
+-- | Run the IO code which syncs the Marconi Utxo indexer with the local running Cardano node.
+runIndexerSyncing
+    :: FilePath
+    -- ^ Marconi indexer database directory
+    -> FilePath
+    -- ^ Local node socket file path
+    -> TMVar UtxoIndexer
+    -> IO ()
+runIndexerSyncing databaseDir nodeSocketPath indexerTVar = do
+    let callbackUtxoIndexer :: UtxoIndexer -> IO ()
+        callbackUtxoIndexer utxoIndexer = atomically $ writeTMVar indexerTVar utxoIndexer
+    (chainPointsToResumeFrom, coordinator) <-
+      startIndexers
+          [ ( utxoWorker callbackUtxoIndexer Nothing
+            -- TODO This hardcoded value needs to be the same value passed through the CLI params in
+            -- the `marconi-chain-index` exe.
```

Comment:
The test seems to already run an indexer, so I don't need to run the marconi-chain-index exe separately, do I?

Reply:
That's true :) In my case, I *did* run a separate `marconi-chain-index` exe to get access to the logging. I'll remove that comment.